src/nti/transactions/loop.py

```diff
@@ -488,23 +512,32 @@ def __handle_generic_exception(self, tx, attempts_remaining, _reraise=six.rerais
             exc_info = None
 
         if self.sleep:
-            _sleep(self.sleep)
+            attempt_num = self.attempts - attempts_remaining
+            backoff = self.random.randint(0, 2 ** attempt_num - 1)
+            self._sleep(self.sleep * backoff) # pylint:disable=too-many-function-args
```

Comment:
One other thought. Do we want a hook to measure and report this (perfmetrics or debug logging)? I could imagine wanting to see metrics around number of retries and amount slept per request. That seems useful to detect potential concurrency related issues?

Reply:
I can see moving averages of total retries being a useful metric. (Ideally you want the average of the average retries per request; I think one of the stat layers can do that correlation?) I'll open a ticket.

I'm less sure about sleep time being a useful metric. We can calculate the expected sleep time given the number of retries, so if we have that we can say what amount of time the system as a whole was spending backing off. I could imagine wanting to see it for an individual request in isolation, especially if we go with the dynamic sleep times, so that can probably be worked into logging. I'll open a ticket.

(Aside: The utility functions in RelStorage for logging durations of functions at different levels might make a good addition to perfmetrics. A slight generalization on them would be useful here. They could be cython compiled to reduce their overhead.)